CLIENT-2476 - Disallow use of add-ops in foreground scans/queries

DomPeliniAerospike opened the pull request:
CLIENT-2476
Disallowed use of operations with foreground queries (foreach and results) Added sufficent testing for each error condition

DomPeliniAerospike committed:
CLIENT-2476

CLIENT-2476
Disallowed use of operations with foreground queries (foreach and results)
Added sufficent testing for each error condition

juliannguyen4 commented on test/new_tests/test_query_apply.py:
```diff
@@ -508,3 +509,57 @@ def _no_items_have_been_applied(self):
         else:
             assert bins["name"] == "no_set_name"
+
+    def test_fail_if_operations_exist_before_apply(self):
```
Shouldn't these tests be checking query.foreach() and query.results()?

DomPeliniAerospike replied:
I included the wrong test cases in this branch, I removed them and added the correct ones, so these comments should not be relevant anymore.

juliannguyen4 commented on test/new_tests/test_query_apply.py:
```diff
@@ -508,3 +509,57 @@ def _no_items_have_been_applied(self):
+            operations.increment("testBinName", 1)
+        ]
+
+        try:
+            query_results = (
+                self.as_connection.query(
+                    "test",
+                    "demo",
+                )
+                .add_ops(ops)
+                .apply(
+                    "query_apply_parameters",
+                    "query_params",
+                    [["age", 5], pickle.dumps({"lary", "quinton", "julie", "mark"})],
+                )
+                .results()
+            )
+            assert "Param error should be thrown" == "This code should not be executed"
+        except e.ParamError:
+            pass
```
1. You should split up this one line into multiple lines, and use `with pytest.raises(e.ParamError)` to ensure that the error happens. If you use `try` and `except`, the error might not be thrown and the test would still pass, which is not correct

```suggestion
      query = self.as_connection.query("test", "demo")
      query.add_ops(ops)
      query.apply(
          "query_apply_parameters",
          "query_params",
          [["age", 5], pickle.dumps({"lary", "quinton", "julie", "mark"})],
      )
      with pytest.raises(e.ParamError) as excinfo:
          query.results()
      assert excinfo.value.msg == "No operations can be used with query.results"
```

DomPeliniAerospike replied:
I included the wrong test cases in this branch, I removed them and added the correct ones, so these comments should not be relevant anymore.

juliannguyen4 commented on test/new_tests/test_query_apply.py:
```diff
@@ -508,3 +509,57 @@ def _no_items_have_been_applied(self):
+        except e.ParamError:
+            pass
+
+    def test_fail_if_UDF_exists_before_add_ops(self):
+        """
+        Invoke query.apply() with a stream udf.
+        arguments contain a serialized set.
+        """
+        ops = [
+            operations.increment("testBinName", 1)
+        ]
+
+        try:
+            query_results = (
+                self.as_connection.query(
+                    "test",
+                    "demo",
+                )
+                .apply(
+                    "query_apply_parameters",
+                    "query_params",
+                    [["age", 5], pickle.dumps({"lary", "quinton", "julie", "mark"})],
+                )
+                .add_ops(ops)
+                .results()
+            )
+            assert "Param error should be thrown" == "This code should not be executed"
+        except e.ParamError:
+            pass
```
I don't think we need a second test case since the first one already tests query.results()
```suggestion
    def test_fail_if_UDF_exists_before_add_ops(self):
        """
        Invoke query.apply() with a stream udf.
        arguments contain a serialized set.
        """
        ops = [
            operations.increment("testBinName", 1)
        ]

        try:
            query_results = (
                self.as_connection.query(
                    "test",
                    "demo",
                )
                .apply(
                    "query_apply_parameters",
                    "query_params",
                    [["age", 5], pickle.dumps({"lary", "quinton", "julie", "mark"})],
                )
                .add_ops(ops)
                .results()
            )
            assert "Param error should be thrown" == "This code should not be executed"
        except e.ParamError:
            pass
```

DomPeliniAerospike replied:
I included the wrong test cases in this branch, I removed them and added the correct ones, so these comments should not be relevant anymore.

DomPeliniAerospike committed:
Disallowed use of ops with scan.results and scan.foreach

Disallowed use of ops with scan.results and scan.foreach
Replaced test cases with the corect test cases.
Fixed lint issuess

juliannguyen4 commented on test/new_tests/test_query.py:
```diff
@@ -807,6 +807,43 @@ def test_query_with_multiple_results_call_on_same_query_object(self):
+
+    def test_query_with_foreach_fail_with_operations(self):
+        """
+        Invoke query() with multple results() call on same query object
```
```suggestion
        Invoke query.foreach() after adding operations
```

juliannguyen4 commented on test/new_tests/test_query.py:
```diff
@@ -807,6 +807,43 @@ def test_query_with_multiple_results_call_on_same_query_object(self):
 
+    def test_query_with_results_fail_with_operations(self):
+        """
+        Invoke query() with multple results() call on same query object
```
```suggestion
        Invoke query.results() after adding operations
```

juliannguyen4 commented on test/new_tests/test_scan.py:
```diff
@@ -467,3 +466,38 @@ def callback(input_tuple):
+
+    def test_scan_with_results_fail_with_operations(self):
+        """
+        Invoke query() with multple results() call on same query object
```
```suggestion
        Invoke scan.results() after adding operations
```

juliannguyen4 commented on test/new_tests/test_scan.py:
```diff
@@ -467,3 +466,38 @@ def callback(input_tuple):
+
+    def test_scan_with_foreach_fail_with_operations(self):
+        """
+        Invoke query() with multple results() call on same query object
```
```suggestion
        Invoke scan.foreach() after adding operations
```

juliannguyen4 commented:
You should also specify in the documentation for query.add_ops() that it should not be used for foreground queries (i.e when calling `query.results()` or `query.foreach()`). Also do the same for scans.

DomPeliniAerospike committed:
Update test/new_tests/test_query.py

Co-authored-by: juliannguyen4 <109386615+juliannguyen4@users.noreply.github.com>

DomPeliniAerospike committed:
Update test/new_tests/test_scan.py

Co-authored-by: juliannguyen4 <109386615+juliannguyen4@users.noreply.github.com>

DomPeliniAerospike committed:
Update test/new_tests/test_query.py

Co-authored-by: juliannguyen4 <109386615+juliannguyen4@users.noreply.github.com>

DomPeliniAerospike commented:
> You should also specify in the documentation for query.add_ops() that it should not be used for foreground queries (i.e when calling `query.results()` or `query.foreach()`). Also do the same for scans.

I will add documentation to explain this and let you know when it has been finished.

DomPeliniAerospike committed:
Update test/new_tests/test_scan.py

Co-authored-by: juliannguyen4 <109386615+juliannguyen4@users.noreply.github.com>

DomPeliniAerospike committed:
Added Documentation Note

juliannguyen4 commented:
Docs preview - https://aerospike-python-client--552.org.readthedocs.build/en/552/scan.html
